### ISSUE TITLE
SSE math in atmos

### ISSUE THREAD

**Issue by kremius** [griefly/griefly#387]
It may give a significant performance rise, but testing is needed.

**Comment by kremius** [comment]
https://github.com/ridiculousfish/libdivide can be helpful

**Comment by kremius** [comment]
http://stackoverflow.com/questions/6996764/fastest-way-to-do-horizontal-float-vector-sum-on-x86
About horizontal sum

**Comment by kremius** [comment]
1. Separate functions for each variation of nearby cells (5 cells, 4 cells, 3 cells, 2 cells, 1 cells)
2. Load all `__m128i` variable to registers and after work only with them.
3. Proper remainder of the division calculation (only from 1 to 5)

**Comment by kremius** [comment]
Approximate upper performance bound for SSE math is 14% (from 21% to 14%) - calculate by replacing `ProcessFiveCells` function by very simple dummy function.
In reality it will be good if cpu use will decrease to 17-18%.
For a while it just does not worth to spend time and efforts on SSE.

**Comment by kremius** [comment]
One good trick will be to switch by possible passability combinations - we have 16 variants for each tile
(fully passable + 4 1 tile impassable + 6 2 tile impassable + 4 3 tile impassable + fully impassable).
It is possible to switch by code paths (easy to implement, but bad for branch prediction) or in some way do branchless version with fetching some precalculated data by index (e.g. a gas value of non-reachable tiles anded with 0 and etc).

**Comment by kremius** [comment]
```
 N
WCE
 S

masks = masks_array[C.passability];
sum_gases = 0;
sum_gases += N.gases & masks.N;
sum_gases += S.gases & masks.S;
sum_gases += E.gases & masks.E;
sum_gases += W.gases & masks.W;

div = popcount(C.passability);
per = sum_gases / div;
leftover = sum_gases - per * div;

all_gases = hor_summ(per);

flow = (all_gases - hor_summ(N.gases)) & masks.N;
N.flows[SOUTH] += flow;
C.flows[NORTH] -= flow;
N.gases -= N.gases & masks.N;
N.gases += per & masks.N;

flow = (all_gases - hor_summ(S.gases)) & masks.S;
S.flows[NORTH] += flow;
C.flows[SOUTH] -= flow;
S.gases -= S.gases & masks.S;
S.gases += per & masks.S;

flow = (all_gases - hor_summ(E.gases)) & masks.E;
E.flows[WEST] += flow;
C.flows[EAST] -= flow;
E.gases -= E.gases & masks.E;
E.gases += per & masks.E;

flow = (all_gases - hor_summ(W.gases)) & masks.W;
W.flows[EAST] += flow;
C.flows[WEST] -= flow;
W.gases -= W.gases & masks.W;
W.gases += per & masks.W;

С.gases = per + leftover;

```

**Comment by kremius** [comment]
https://www.cs.uaf.edu/2011/fall/cs441/lecture/09_27_branch_and_SIMD.html

However, seems like SIMD instructions are not heavily affected by branch prediction, so it might be reasonable to switch by code paths instead.
Anyway, testing is needed.

**Comment by kremius** [comment]
https://github.com/kremius/sse-branch-prediction
tl;dr:
simple SSE will speed up a little bit, but probably bottleneck will be not here, but in the memory loads & the atmos objects movement processing.
